book/visual-effects.md

```diff
@@ -1379,43 +1303,62 @@ def parse_blend_mode(blend_mode_str):
         return skia.BlendMode.kSrcOver
 ```
 
-This makes adding support for blend modes to our browser as simple as
-passing the `BlendMode` parameter to the `Paint` object:
+We can make a corresponding draw command using `saveLayer`:
+
+``` {.python}
+class Blend:
+    def __init__(self, blend_mode, children):
+        self.blend_mode = parse_blend_mode(blend_mode)
+        self.children = children
+        self.rect = skia.Rect.MakeEmpty()
+        for cmd in self.children:
+            self.rect.join(cmd.rect)
+
+    def execute(self, canvas):
+        paint = skia.Paint(BlendMode=self.blend_mode)
+        canvas.saveLayer(paint=paint)
+        for cmd in self.children:
+            cmd.execute(canvas)
+        canvas.restore()
+```
+
+Then we just add a `Blend` operation to `paint_visual_effects`:
 
 ``` {.python expected=False}
 def paint_visual_effects(node, cmds, rect):
     # ...
-    blend_mode = parse_blend_mode(node.style.get("mix-blend-mode"))
+    blend_mode = node.style.get("mix-blend-mode")
     
     return [
-        SaveLayer(skia.Paint(BlendMode=blend_mode), [
-            SaveLayer(skia.Paint(Alphaf=opacity), cmds),
+        Blend(blend_mode, [
+            Alpha(opacity, cmds),
         ]),
     ]
 ```
 
 Note the order of operations here: we _first_ apply transparency, and
-_then_ blend the result into the rest of the page. If we switched the
-two `SaveLayer` calls, so that we first applied blending, there
-wouldn't be anything to blend it into!
```

Comment:
Is this last sentence wrong, or do you think it doesn't add value?

Reply:
Oh yeah great Q. I think the original explanation was correct but pretty tough to understand. The new version is totally different but maybe also correct and easy to understand. Willing to hear otherwise.